tsh PIV login integration

Joerger opened the pull request:
Part 2 for implementing https://github.com/gravitational/teleport/pull/14774

Changes:
 - Adds `YubikeyPrivateKey`, which implements `keys.PrivateKey` while interfacing through a Yubikey PIV module using `piv-go`
 - Adds `PIV_LOGIN` env var to trigger PIV login. This will be removed after https://github.com/gravitational/teleport/pull/15874, which will automatically determine when PIV login is needed.
 - Add logic to reuse PIV keys set up by Teleport client
 - Move retry/backoff utils to `/api/utils/retryutils` for use in Yubikey open logic.

Notes:
 - Connections to yubikey PIV are created ad hoc when `YubikeyPrivateKey.Sign` is called. Since multiple connections are created and closed for each `tsh` request and PIV connections take some additional time to fully close on the PIV side, we are using a short retry loop to connect. Additionally, since we reopen the PIV connection several times per request, this adds a significant (1-5 second) latency. In a future PR, we can replace this primitive retry/reopen logic with singleton connection logic shared across an entire `tsh` call, since we only ever connect to a single yubikey.
 - Agent forwarding functionality does not work with yubikey login, since `YubikeyPrivateKey` isn't supported by `keys.AsAgentKey`. This breaks some functionality including `tsh proxy ssh`, server sessions in proxy recording mode, `ssh -A` and `tsh ssh -A`.
 - Kubernetes integration is also not supported with yubikey login
 - `tsh db connect` will not work with yubikey login, but `tsh proxy db` does work and can be used as a workaround.

Joerger committed:
Add Yubikey PrivateKey implementation for use by Teleport clients.

  - Add yubikey login logic, reusing previously stored private keys.

  - Fix identity file decoding with PIV keys, which sign ecdsa certificates.

  - Add libpcsclite-dev pre-req for building on linux.

  - Remove unnecessary keys.Signer interface and move its functionality to keys.PrivateKey.

  - Add new api/utils/retryutils package.

espadolini commented on api/utils/keys/yubikey.go:
```diff
@@ -0,0 +1,407 @@
+		return strings.Contains(err.Error(), retryError)
+	}
+
+	var maxRetries int = 100
```
This can block for quite a long time and there's no way to cancel it either - is it worth it to pass along some context or is this only ever going to be used by tsh/Connect so it doesn't matter?

Even in that case, blocking for up to 10 seconds before closing the program seems a bit of a UI blunder.

Joerger replied:
> This can block for quite a long time and there's no way to cancel it either - is it worth it to pass along some context or is this only ever going to be used by tsh/Connect so it doesn't matter?

This would require adding `ctx` to `crypto.Signer.Sign`, which is not possible while maintaining functionality as a `crypto.Signer`. I think the only solution we can introduce is to reduce the retry limit to (ideally) under 100 milliseconds (which we can't do until the below issue is address)

> Even in that case, blocking for up to 10 seconds before closing the program seems a bit of a UI blunder.

Retries primarily occur when a PIV connection is opened during or immediately after a previous connection. A short backoff is sufficient for `tsh` which execute sequentially, but Teleport Connect requires a longer backoff since it opens the PIV private key countless times on startup.

Since PIV devices only allow a single connection at a time, the primary way we can improve this is by reusing connections. Since any Teleport process can open a PIV connection at any time to access it's associated private key, we would need to maintain a global connection pool with cached clients that are closed after a delay. 

I'm not sure of the best way to implement this and am concerned that it will block the PR, so I was planning on making this improvement in a follow up PR (after https://github.com/gravitational/teleport/pull/15336 which introduces the remaining core functionality).

espadolini replied:
And we'd need some retry logic for the first opening of the key anyway, unless we're fine with just erroring out and exiting 😢

Joerger replied:
Yes :( We could potentially open the yubikey ahead of time under a `ctx`, but managing the connection and closing it when it's not being used would become a struggle. For example, while the PIV connection is held open, the yubikey cannot accept MFA requests.

espadolini commented on api/utils/keys/yubikey.go:
```diff
@@ -0,0 +1,407 @@
+const certOrgName = "teleport"
+
+func selfSignedTeleportClientCertificate(priv crypto.PrivateKey, pub crypto.PublicKey) (*x509.Certificate, error) {
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
```
```suggestion
	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128) // see crypto/tls/generate_cert.go
```

espadolini commented on api/utils/keys/yubikey.go:
```diff
@@ -0,0 +1,407 @@
+
+	// Verify that the slot's certs have the same public key, otherwise the key
+	// may have been generated by a non-teleport client.
+	if !cert.PublicKey.(interface{ Equal(x crypto.PublicKey) bool }).Equal(slotCert.PublicKey) {
```
Do we have a hard guarantee that the `PublicKey` for a `x509.Certificate` can only ever be a valid (the interface as described by the comment for) `crypto.PublicKey`?

Joerger replied:
Actually no, because looking through the code path it could potentially be a `*dsa.PublicKey` which doesn't implement `Equal`. I'll add the check, thanks.

Joerger commented on api/utils/keys/privatekey.go:
```diff
@@ -36,9 +36,10 @@ import (
+	pkcs1PrivateKeyType      = "RSA PRIVATE KEY"
+	ecPrivateKeyType         = "EC PRIVATE KEY"
+	pkcs8PrivateKeyType      = "PRIVATE KEY"
+	pivYubikeyPrivateKeyType = "PIV YUBIKEY PRIVATE KEY"
```
The serial number serialized in the PEM block is specific to hardware key manufacturue, so we either need to specify by block type - `PIV YUBIKEY PRIVATE KEY` - or with an additional serialized field - `type: yubikey`. I don't have a strong preference either way, I can see arguments for both.

Joerger commented on api/utils/keys/yubikey.go:
```diff
@@ -0,0 +1,407 @@
+	"time"
+
+	"github.com/go-piv/piv-go/piv"
+	"github.com/gravitational/teleport/api"
```
To think I've been doing this manually this whole time...

espadolini commented on api/utils/keys/yubikey.go:
```diff
@@ -0,0 +1,384 @@
+		}
+	}
+
+	if len(yubiKeyCards) == 0 {
+		return nil, trace.NotFound("no yubiKey devices found")
+	}
+
```
"No cards connected" can be signaled by returning zero cards - otherwise returning `[]string{}, nil` would be possible but invalid.

```suggestion
```

espadolini commented on api/utils/keys/yubikey.go:
```diff
@@ -0,0 +1,384 @@
+		return nil, trace.Wrap(err)
+	}
+
+	y, err := findYubiKey(keyData.SerialNumber)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	pivSlot, err := parsePIVSlot(keyData.SlotKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```
Swap the order two checks, do the one that's just in-memory first.

espadolini commented on api/utils/keys/yubikey.go:
```diff
@@ -0,0 +1,384 @@
+		return nil, trace.Wrap(err)
+	}
+
+	// Store an unsigned certificate to mark this slot as used by tsh.
```
```suggestion
	// Store a self-signed certificate to mark this slot as used by tsh.
```

espadolini commented on api/utils/keys/yubikey.go:
```diff
@@ -0,0 +1,384 @@
+	cert := &x509.Certificate{
+		SerialNumber: serialNumber,
+		PublicKey:    pub,
+		Subject: pkix.Name{
```
Do we want to store serial and slot in the subject for later debugging, or is it just not necessary?

Joerger replied:
I don't think so, that data should already be known if you're able to find the certificate in the first place

espadolini commented on api/utils/keys/yubikey.go:
```diff
@@ -0,0 +1,384 @@
+// matching the given touch requirement. This private key will either be newly generated or previously
+// generated by a Teleport client and reused.
+func GetOrGenerateYubiKeyPrivateKey(touchRequired bool) (*PrivateKey, error) {
+	// Use the first yubiKey we find.
+	y, err := findYubiKey(0)
```
Do we plan on letting the user select keys in the future in some way?

Joerger replied:
Probably not, as this idea was rejected during RFD review. The current goal is to enable PIV login without any additional flags, options, etc.

espadolini reviewed (approved):
Are we going to fork `x/crypto/ssh/agent` to support generic `ssh.Signer`s?

Joerger commented:
> Are we going to fork `x/crypto/ssh/agent` to support generic `ssh.Signer`s?

Most likely yes, as that will probably be necessary to fix OpenSSH agent forwarding integration. However, we may be able to avoid this temporarily by fixing the Teleport local agent without touching `x/crypto/ssh/agent`, or by implementing another workaround, such as introducing PIV key logic directly into proxy commands like `tsh proxy ssh`. This is planned to be investigated in depth in a follow up PR.

espadolini commented on api/identityfile/identityfile.go:
```diff
@@ -301,3 +302,25 @@ func decodeIdentityFile(idFile io.Reader) (*IdentityFile, error) {
+	ssh.CertAlgoSKECDSA256v01,
+	ssh.CertAlgoED25519v01,
+	ssh.CertAlgoSKED25519v01,
+	ssh.CertAlgoRSASHA256v01,
+	ssh.CertAlgoRSASHA512v01,
```
Looks like these two should never actually appear as on-disk certs, FYI.

Should we check for `-cert-v.*@openssh.com` instead?

Joerger replied:
Sure, I've updated it to use a regex

Joerger commented:
@tcsc Can you review when you get a chance?

Also, CI is failing and I don't understand why. I think I'm missing something, so any help/pointers would be very helpful. From what I can tell, just adding `apt-get libpsclite-dev` to the `build.assets/dockerfile` should be sufficent to make [go-piv](https://github.com/go-piv/piv-go#installation) build properly, but unit and integration tests are failing with:

```
CGO_ENABLED=1 CGO_LDFLAGS="-Wl,-Bstatic -lbpf -lelf -lz -Wl,-Bdynamic -Wl,--as-needed" go test -json -run "^TestRoot" github.com/gravitational/teleport/integration github.com/gravitational/teleport/integration/helpers github.com/gravitational/teleport/integration/hsm -v -race \
	| tee /workspace/test-logs/integration-root.json \
	| /workspace/build.assets/tooling/bin/render-tests -report-by test
# pkg-config --cflags  -- libpcsclite
Package libpcsclite was not found in the pkg-config search path.
Perhaps you should add the directory containing `libpcsclite.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libpcsclite' found
```

Joerger commented:
@espadolini I've made some significant changes, specifically to add the `libpcsclite` build flag needed to complie `piv-go`, so you might want to re-review.

espadolini commented on Makefile:
```diff
@@ -182,6 +182,18 @@ TOUCHID_MESSAGE := with Touch ID
+ifeq ("$(LIBPCSCLITE)", "yes")
+LIBPCSCLITE_MESSAGE := with libpcsclite
+LIBPCSCLITE_BUILD_TAG := libpcsclite
+else ifeq ("$(shell pkg-config libpcsclite 2>/dev/null; echo $$?)", "0")
```
Can we add a way to opt-out of this? We don't do any autodetection for `libfido`, and there's at least ways to force the Rust builds and libbpf off.

espadolini commented on Makefile:
```diff
@@ -231,21 +243,21 @@ all: version
 .PHONY: $(BUILDDIR)/teleport
 $(BUILDDIR)/teleport: ensure-webassets bpf-bytecode rdpclient
-	GOOS=$(OS) GOARCH=$(ARCH) $(CGOFLAG) go build -tags "$(PAM_TAG) $(FIPS_TAG) $(BPF_TAG) $(WEBASSETS_TAG) $(RDPCLIENT_TAG)" -o $(BUILDDIR)/teleport $(BUILDFLAGS) ./tool/teleport
+	GOOS=$(OS) GOARCH=$(ARCH) $(CGOFLAG) go build -tags "$(PAM_TAG) $(FIPS_TAG) $(BPF_TAG) $(WEBASSETS_TAG) $(RDPCLIENT_TAG) $(LIBPCSCLITE_BUILD_TAG)" -o $(BUILDDIR)/teleport $(BUILDFLAGS) ./tool/teleport
```
Is this here in `teleport` because we're eventually going to need the PIV library on the server side?

Joerger replied:
On the server side, we use `go-piv`'s attestation logic.

espadolini commented on api/identityfile/identityfile.go:
```diff
@@ -261,7 +262,7 @@ func decodeIdentityFile(idFile io.Reader) (*IdentityFile, error) {
 	for scanln() {
 		switch {
-		case hasPrefix("ssh"):
+		case isSSHCert(line):
```
Thinking about this again, would it be too slow to `ssh.ParseAuthorizedKeys` and check that the public key we got is a `*ssh.Certificate`?

Joerger replied:
I think the performance difference would be negligible. Honestly, `decodeIdentityFile` should probably just fully parse the entire file into it's actual `ssh.Certificate`, `tls.Certificate`, `keys.PrivateKey`, etc, since they will need to be parsed (probably multiple times) afterwards.

Joerger replied:
On second thought, I think this is out of scope, since currently the decode function is is just meant to parse and organize the identity file parts, rather than verify the contents. For example, if there is an ssh cert present, but it's invalid, decode should not return "ssh cert missing". Instead, we should parse it after decoding.

espadolini commented on dronegen/mac.go:
```diff
@@ -521,7 +521,7 @@ func darwinBuildCommands(toolchainConfig toolchainConfig, artifactConfig darwinA
 			`build.assets/build-fido2-macos.sh build`,
 			`export PKG_CONFIG_PATH="$(build.assets/build-fido2-macos.sh pkg_config_path)"`,
-			`make clean release OS=$OS ARCH=$ARCH FIDO2=yes TOUCHID=yes`,
+			`make clean release OS=$OS ARCH=$ARCH FIDO2=yes TOUCHID=yes LIBPCSCLITE=yes`,
```
Is libpcsclite preinstalled on macOS?

Joerger replied:
That's what I thought, but actually it's just not required for macOS or Windows to compile `piv-go`. I removed the tag from here and added the `!linux` build tag to `keys/yubikey.go`. 

Let me know if I'm missing any other important build steps, I haven't touched the new build flow much.

tcsc commented:
You will probably need to add `libpsclite` to the buildbox in a separate PR. The test images are only updated on merge, so adding it inside this PR means that you're still testing on the old image.

Here's one I've prepared earlier: https://github.com/gravitational/teleport/pull/16083

Joerger commented on api/utils/keys/yubikey.go:
```diff
@@ -0,0 +1,383 @@
+			return nil, trace.Wrap(err)
+		}
+
+		time.Sleep(time.Millisecond * 100)
```
1. You're probably right, I'll move more Jitter/Retry logic from lib to api. Looks like some jitter logic was already moved.
2. Sure, good idea
3. We'll have to add ctx as an argument for all of the key parsing functions, but you're probably right that it's necessary. 

FYI: The yubikey tests require an actual Yubikey be attached, so `yubikey_test.go` won't be running in CI. Existing tests continue to use RSA private keys, since there isn't a yubikey PIV test module (that I could find).

Joerger replied:
I went ahead and added in proper retry/context logic, but I decided not to pipe the context through all yubikey code for now to prevent over sizing this PR.

The connection logic is currently a work in progress, and I'm planning to create a follow up PR to manage open PIV connections instead of creating connections ad-hoc.

Joerger commented on Makefile:
```diff
@@ -191,7 +200,7 @@ endif
 # and tsh.
 BINARIES=$(BUILDDIR)/teleport $(BUILDDIR)/tctl $(BUILDDIR)/tsh $(BUILDDIR)/tbot
-RELEASE_MESSAGE := "Building with GOOS=$(OS) GOARCH=$(ARCH) REPRODUCIBLE=$(REPRODUCIBLE) and $(PAM_MESSAGE) and $(FIPS_MESSAGE) and $(BPF_MESSAGE) and $(RDPCLIENT_MESSAGE) and $(LIBFIDO2_MESSAGE) and $(TOUCHID_MESSAGE)."
+RELEASE_MESSAGE := "Building with GOOS=$(OS) GOARCH=$(ARCH) REPRODUCIBLE=$(REPRODUCIBLE) and $(PAM_MESSAGE) and $(FIPS_MESSAGE) and $(BPF_MESSAGE) and $(RDPCLIENT_MESSAGE) and $(LIBFIDO2_MESSAGE) and $(TOUCHID_MESSAGE) and $(LIBPCSCLITE_MESSAGE)."
```
Fixed in https://github.com/gravitational/teleport/pull/16424/commits/7937f9824f55858cbd20c5bd4e2b0962e3c58e0c (https://github.com/gravitational/teleport/pull/16424)

jakule commented on Makefile:
```diff
@@ -278,7 +278,7 @@ $(BUILDDIR)/tsh:
 .PHONY: $(BUILDDIR)/tbot
 $(BUILDDIR)/tbot:
-	GOOS=$(OS) GOARCH=$(ARCH) $(CGOFLAG) go build -tags "$(FIPS_TAG)" -o $(BUILDDIR)/tbot $(BUILDFLAGS) ./tool/tbot
+	GOOS=$(OS) GOARCH=$(ARCH) $(CGOFLAG) go build -tags "$(FIPS_TAG) $(LIBPCSCLITE_BUILD_TAG)" -o $(BUILDDIR)/tbot $(BUILDFLAGS) ./tool/tbot
```
Why `tbot` needs that library? Shouldn't it be just a `tsh` thing?

Joerger committed:
Resolve comments.

Joerger committed:
Update e ref to master.

Enable go-piv only when libpcsclite build tag is included explicitly.

Fix ssh cert regex.

Joerger commented on api/identityfile/identityfile.go:
```diff
@@ -301,3 +302,15 @@
+
+// OpenSSH cert types look like "<key-type>-cert-v<version>@openssh.com".
+// Currently, we only use "ssh-rsa-cert-v01@openssh.com" & "ecdsa-sha2-nistp256-cert-v01@openssh.com".
+var sshCertTypeRegex = regexp.MustCompile(`^[a-z0-9\-]+-cert-v[0-9]{2}@openssh.com$`)
```
fixed
